torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,362 @@
+}
+
+at::Tensor allocate_all_gather_output(
+    const at::Tensor& input,
+    int64_t group_size) {
+  auto output_size = input.sizes().vec();
+  output_size[0] *= group_size;
+  return at::empty(
+      output_size,
+      at::TensorOptions().dtype(input.dtype()).device(input.device()));
+}
+
+std::vector<at::Tensor> all_gather_into_tensor_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const int64_t group_size,
+    const std::string& tag) {
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(allocate_all_gather_output(tensor, group_size));
+  }
+
+  auto pg = c10d_functional::resolve_process_group(tag);
+  auto work = pg->allgather_into_tensor_coalesced(
+      outputs, const_cast<std::vector<at::Tensor>&>(inputs));
+  for (const auto& tensor : outputs) {
+    WorkRegistry::get().register_work(tensor, work);
+  }
+  return outputs;
+}
+
+at::Tensor all_gather_into_tensor(
+    const at::Tensor& input,
+    const int64_t group_size,
+    const std::string& tag) {
+  std::vector<at::Tensor> inputs{input};
+  return all_gather_into_tensor_coalesced(inputs, group_size, tag)[0];
+}
+
+at::Tensor allocate_reduce_scatter_output(
+    const at::Tensor& input,
+    const int64_t group_size) {
+  auto output_size = input.sizes().vec();
+  if (output_size[0] % group_size != 0) {
+    LOG(WARNING) << "The first dimension of the reduce_scatter input ("
+                 << output_size[0] << ") is not divisible by the group size ("
+                 << group_size << ").";
+  }
+  output_size[0] /= group_size;
+  return at::empty(
+      output_size,
+      at::TensorOptions().dtype(input.dtype()).device(input.device()));
+}
+
+std::vector<at::Tensor> reduce_scatter_tensor_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const std::string& reduce_op,
+    const int64_t group_size,
+    const std::string& tag) {
+  c10d::ReduceScatterOptions opts;
+  opts.reduceOp = to_reduce_op(reduce_op);
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(allocate_reduce_scatter_output(tensor, group_size));
+  }
+
+  auto pg = c10d_functional::resolve_process_group(tag);
+  auto work = pg->reduce_scatter_tensor_coalesced(
+      outputs, const_cast<std::vector<at::Tensor>&>(inputs), opts);
+  for (const auto& tensor : outputs) {
+    WorkRegistry::get().register_work(tensor, work);
+  }
+  return outputs;
+}
+
+at::Tensor reduce_scatter_tensor(
+    at::Tensor input,
+    const std::string& reduce_op,
+    const int64_t group_size,
+    const std::string& tag) {
+  std::vector<at::Tensor> inputs{input};
+  return reduce_scatter_tensor_coalesced(inputs, reduce_op, group_size, tag)[0];
+}
+
+at::Tensor wait_tensor(const at::Tensor& tensor) {
+  WorkRegistry::get().pop_and_wait_work(tensor);
+  return tensor;
+}
+
+} // namespace
+
+TORCH_LIBRARY(_c10d_functional, m) {
+  m.def(
+      "all_reduce(Tensor input, str reduce_op, str tag) -> Tensor",
```

Comment:
"tag" is a bit hard for user to understand, wondering if we should do `group_name` instead?

Reply:
Sounds good!